Userland/Libraries/LibJS/Tests/builtins/RegExp/RegExp.legacy.js

```diff
@@ -0,0 +1,160 @@
+    expect(RegExp['$&']).toBe('123.456');
+    expect(RegExp.lastParen).toBe('456');
+    expect(RegExp['$+']).toBe('456');
+    expect(RegExp.leftContext).toBe('abc');
+    expect(RegExp['$`']).toBe('abc');
+    expect(RegExp.rightContext).toBe('def');
+    expect(RegExp["$'"]).toBe('def');
+
+    expect(RegExp.$1).toBe('123.456');
+    expect(RegExp.$2).toBe('123');
+    expect(RegExp.$3).toBe('456');
+    expect(RegExp.$4).toBe('');
+    expect(RegExp.$5).toBe('');
+    expect(RegExp.$6).toBe('');
+    expect(RegExp.$7).toBe('');
+    expect(RegExp.$8).toBe('');
+    expect(RegExp.$9).toBe('');
+});
+
+test('test should all behave as if exec were called.', () => {
+    var re = /((\d+)\.(\d+))/;
+    var s = 'ghi789.012jkl';
+    re.test(s);
+
+    expect(RegExp.input).toBe(s);
+    expect(RegExp['$_']).toBe(s);
+    expect(RegExp.lastMatch).toBe('789.012');
+    expect(RegExp['$&']).toBe('789.012');
+    expect(RegExp.lastParen).toBe('012');
+    expect(RegExp['$+']).toBe('012');
+    expect(RegExp.leftContext).toBe('ghi');
+    expect(RegExp['$`']).toBe('ghi');
+    expect(RegExp.rightContext).toBe('jkl');
+    expect(RegExp["$'"]).toBe('jkl');
+
+    expect(RegExp.$1).toBe('789.012');
+    expect(RegExp.$2).toBe('789');
+    expect(RegExp.$3).toBe('012');
+    expect(RegExp.$4).toBe('');
+    expect(RegExp.$5).toBe('');
+    expect(RegExp.$6).toBe('');
+    expect(RegExp.$7).toBe('');
+    expect(RegExp.$8).toBe('');
+    expect(RegExp.$9).toBe('');
+});
+
+test('replace must interleave matching and replacing when a global regexp is matched and replaced with the result of a function', () => {
+    var re = /(.)/g;
+    function f() { return RegExp.$1; };
+    expect('abcd'.replace(re, f)).toBe('dddd');
+});
+
+test('the last parenthesis didn\'t match.', () => {
+    expect(/foo(?:a(x))?/.exec("foobx")).toEqual(["foo", undefined]);
+    expect(RegExp.lastParen).toBe('');
+});
+
+test('$1 to $9', () => {
+    for (var i = 1; i <= 9; i++) {
+        var haystack = "foo";
+        var re_text = "^foo";
+        for (var j = 0; j < i - 1; j++) {
+            haystack += "x";
+            re_text += "(x)";
+        }
+        re_text += "(?:a(x))?";
+        haystack += "bx";
+        var re = new RegExp(re_text);
+        expect(re.test(haystack)).toBeTrue();
+        for (var j = 1; j < i - 1; j++) {
+            expect(RegExp['$' + j]).toBe("x");
+        }
+        expect(RegExp['$' + (i)]).toBe("");
+    }
+});
+
+test('save the correct string as the last subject', () => {
+    var foo = "lsdfj sldkfj sdklfj læsdfjl sdkfjlsdk fjsdl fjsdljskdj flsj flsdkj flskd regexp: /foobar/\nldkfj sdlkfj sdkl";
+    expect(/^([a-z]+): (.*)/.test(foo.substring(foo.indexOf("regexp:")))).toBeTrue();
+    expect(RegExp.$1).toBe("regexp");
+});
+
+test('calling with no argument is the same as calling with undefined.', () => {
+    expect(/^undefined$/.test()).toBeTrue();
+    expect(/^undefined$/.exec()).toEqual(["undefined"]);
+})
```

Comment:
Maybe we can also test what the values are without having ran any RegExp calls?
That might need to be it's own file though because this is global state